Alternative way to address consistency issues mentioned in #491 and #493

pjdietz opened the pull request:
Another variant of PR #491 and #493 with specific return types.

This PR suggests the following changes:

- `getHeader($name)` takes the place of the current `getHeaderLines($name)` and returns an array of strings (even with only one).
- `getHeaders()` is unchanged and returns an array with arrays of strings for values.
- `getHeaderLine($name)` takes the place of the current `getHeader($name)` and returns a (concatenated) string.
- `getHeaderLines()` returns an array with strings as values.

This provides a predictable interface where the user can know the return type in advance and won't need to test whether the method returned a string or array.

It also adds consistency where both:

```
$message->getHeaders()[$k] === $message->getHeader($k)
```

And

```
$message->getHeaderLines()[$k] === $message->getHeaderLine($k)
```

webmozart commented:
:+1:

weierophinney commented:
@codeguy and @mtdowling — I think it looks good, but would like your opinion as well. I think this clarifies the method names and intentions quite well. It *does* represent a change, however, and as two others who are currently coding implementations, want your feedback.

Thanks!

codeguy commented:
:+1:

mtdowling commented on proposed/http-message.md:
```diff
@@ -383,6 +383,18 @@ interface MessageInterface
+     * Retrieves a header by the given case-insensitive name as an array of
+     * strings.
+     *
+     * If the header did not appear in the message, this method should return an
```
If the header does not appear in the message, this method MUST return an empty array.

weierophinney commented:
In talking with @mtdowling just now, we want to point out that the following would not be true:

```php
$message->getHeader($k) === $message->getHeaders()[$k];
```

**unless** the header key `$k` matches the casing used to inject the header. (Not that this would be a show-stopper, but it should be noted.)

mtdowling commented on proposed/http-message.md:
```diff
@@ -542,7 +558,7 @@ interface RequestInterface extends MessageInterface
      * @see UriInterface::getHost()
      * @param string $name Case-insensitive header field name.
-     * @return string
+     * @return string[]
```
Can you add a description of the return type? e.g., "Returns an array of strings"

weierophinney replied:
Seconded. :)

mtdowling commented on proposed/http-message.md:
```diff
@@ -550,7 +566,7 @@ interface RequestInterface extends MessageInterface
      * behavior.
      *
-     * Retrieves a header by the given case-insensitive name as an array of strings.
+     * Retrieves a header by the given case-insensitive name as a string.
```
It might be good to repeat some of the verbiage in getHeaderLines() that describes the return value (a concatenated string, yadda, yadda...)

mtdowling commented on proposed/http-message.md:
```diff
@@ -399,18 +411,22 @@ interface MessageInterface
 
     /**
-     * Retrieves a header by the given case-insensitive name as an array of strings.
+     * Retrieves all message headers.
```
Maybe something more explicit to help people understand why this method exists: Retrieves all message headers as an array of header name to string values.

pjdietz commented:
@weierophinney Ah, good catch. Still, it's closer. Also not true if `$k` did not appear since `getHeader($k)` will be an empty array and `getHeaders()[$k]` will be an undefined index.

weierophinney commented:
@pjdietz Can you address the more targetted feedback from @mtdowling? I think with those changes, we can likely merge.

Thanks!

mtdowling commented:
This looks good.

I wonder if there's a good reason to have `getHeaderLines()` though. `getHeaderLine($k)` in this PR is useful because it concatenates values together into a string case-insensitively, allowing easy access to headers across the whole message case-insensitively. `getHeaderLines()` isn't that useful because it only concats values together of the same case-insensitive name. It's far less useful unless we complicate the method a bit... For example: returns a normalized list of header names to values (where normalized means it flattens out case-insensitivity across all of the headers). I don't know if it would be worth it at that point.

pjdietz commented:
@weierophinney and @mtdowling will do!

weierophinney commented:
@mtdowling I've been unsure about the plural form for retrieving string values, too. I'm only entertaining it because we've had three separate PRs now trying to make for symmetric methods for singular/plural forms.

pjdietz commented:
I've added the clarification @mtdowling suggested in a new commit. I'm not sure how useful `getHeaderLines()` is either. I liked the direction @pmjones was going with #493 but wanted to avoid having a method return strings for some headers and arrays for others.

mtdowling commented:
I like this PR the most out of all of the current PR's on the table. I do think that since none of us see the value in `getHeaderLines()` that it should be removed. I think that change would make this PR stand out even more as the best approach.

pmjones commented:
Not "none." ;-) Example forthcoming.

pmjones commented:
To wit:

```php
<?php
    foreach ($message->getHeaderLines() as $line) {
        header($line);
    }
```

wolfy-j commented:
This is incorrect example, $line will contain header value but not name, plus it will corrupt such headers as Set-Cookie and etc.

pjdietz commented:
@pmjones That's a more useful getHeaderLines() than in this PR. I missed what you had done for that method in #493.
```
   /**
     * Retrieves all header lines as array of strings.
     *
     * If there are no headers, this method should return an empty array.
     *
     * @return string[]
     */
    public function getHeaderLines();
```

weierophinney commented:
@pmjones and @pjdietz: That's not a great example. First, it should likely be this:

```php
foreach ($message->getHeaderLines() as $header => $value) {
    header(sprintf('%s: %s', $header, $value));
}
```

Second, it will break on some headers, as not all headers can be safely comma-concatenated. The specification indicates that if a header can allow multiple values, comma-concatenation or multiple headers are allowed, but does not specify that they MUST allow comma-concatenation. In particular, `Set-Cookie` and `Cookie` famously do follow that recommendation (they use `;` as a delimiter). This is why @mtdowling and I have been arguing it's a useless method.

pjdietz commented:
Although, yeah, cookies and such won't work.

pmjones commented:
@pjdietz If the primary difference between #493 and #494 is the standardization of returning arrays, that's something I can modify in #493 very easily.

pmjones commented:
@weierophinney Noted -- so then how does one get a list of all header names that are in the message? `$names = array_keys($message->getHeaders();` and then `foreach ($names as $name) { header(sprintf('%s: %s', $name, $message->getHeaderLine($name))); }` ?

pmjones commented:
And honestly I would expect implementations to do the sane thing and have `getHeaderLines()` refer back to `getHeaderLine()` internally -- but perhaps that is naively optimistic on my part.

wolfy-j commented:
@pmjones you can use `getHeaders()` directly to get header name and it's all lines, @weierophinney has great example in his library.

webmozart commented:
@weierophinney Just for the sake of completeness, could you give some examples where comma-concatenation is not valid?

pmjones commented:
@wolfy-j In this and in #493, `getHeaders()` returns an array of values, not a string and not lines.

weierophinney commented:
@webmozart Already did: `Cookie` and `Set-Cookie`.

wolfy-j commented:
@pmjones, i think #493 does change `getHeaders()` behaviour, this PR does not:
> getHeaders() is unchanged and returns an array with arrays of strings for values.

pmjones commented:
Cf. the newly modified #493; arrays are now returned consistently.

weierophinney commented:
> @weierophinney Noted -- so then how does one get a list of all header names that are in the message? $names = array_keys($message->getHeaders(); and then foreach ($names as $name) { header(sprintf('%s: %s', $name, $message->getHeaderLine($name))); } ?

@pmjones I'm not even sure what you're trying to do in that example. `getHeaderLines()` as proposed here returns key/value pairs, where the key is the header name, the value the concatenated string. That's why I suggested:

```php
foreach ($message->getHeaderLines() as $header => $value) {
    // ...
}
```

But yes: you would get the list of headers that are currently present using either `array_keys($message->getHeaderLines())` or `array_keys($message->getHeaders()`, and each would also return the header names in the case with which they were registered.

pmjones commented:
@wolfy-j : Ah so! In which case, I recommend against this PR. I'd like to see something that returns the header values as set, *and* something that returns lines.  In #493, getHeader() and getHeaders() return the array of values; getHeaderLine() returns a single header line concatenated, and getHeaderLines() returns an array of those lines.

wolfy-j commented:
@pmjones, so we ended with decision `getHeader()` vs `getHeaderLine()`, there is no other differences in this PRs.

pjdietz commented:
What's the use case for concatenation? Which headers would not work if `getHeaderLines()` returned an array of strings with an item for each header value? For example, if there are three "Set-cookie" values, the method would return:

```
[
    "Set-cookie: ...", 
    "Set-cookie: ...", 
    "Set-cookie: ...", 
    "X-foo: bar"
]
```

pmjones commented:
Apparently some folks like the idea of concatenated return values. #493 takes that into account, while also allowing access to the array of values as set.

weierophinney commented:
@pmjones See my comment on your own PR; I see it exactly the opposite. What this PR does is clarify the "Line" verbiage, which, semantically, implies strings, but is currently (with `getHeaderLines()`) returning an array. This PR changes the interface to return what you would expect based on the method name.

The only part where I see that as maybe *not* being the case is `getHeader()`, which would now return an array. However, since it is mirrored by `getHeaders()`, and returns the same structure, this makes sense.

The bigger question is whether we need both `getHeaderLine()` and `getHeaderLines()`. As @mtdowling and I have argued, the latter seems like a relatively dangerous idea, as RFC 7230 does not require that a comma be used for concatenation. We have stipulated that in our string versions for consistency and to ease the burden on implementors (as custom header values may have their own requirements, too); as such, having an array of key/string pairs is not terribly useful.

weierophinney commented:
@pjdietz What you just described is the current behavior of `getHeaderLines($name)`… Bringing us full-circle back to the current implementation.

wolfy-j commented:
@pjdietz, i can't see it specified in your original post, even more:
> `$message->getHeaderLines()[$k] === $message->getHeaderLine($k)`

pjdietz commented:
@weierophinney , not quite. What I'm thinking is `getHeaderLines()` with no argument returns the full "{name}: {value}" for every header. Something you could use with this:

```
    foreach ($message->getHeaderLines() as $line) {
        header($line);
    }
```

pjdietz commented:
But where no concatenation takes place.

pjdietz commented:
@wolfy-j Sorry for the confusion. That's not in this PR. Just throwing that out as an idea.

webmozart commented:
@pjdietz I think this method is too specific. You can easily solve this use case with `getHeaders()`.

I think that `getHeaderLines()` can safely be removed then. Apart from that, I like this PR.

wolfy-j commented:
@webmozart :+1:

weierophinney commented:
@pjdietz Eww. No. Again, not a sane solution, due to the problem of concatenation. The concatenation issue is why `getHeaders()` returns the structure it does. You can see how that is consumed here:

- https://github.com/phly/http/blob/master/src/Server.php#L205-L216

Let's go with the following:

- `getHeader($name)` should return the array of values, and an empty array if no values are present or header is missing.
- `getHeaders()` should remain unchanged.
- Add `getHeaderLine($name)`, which would provide a comma-concatenated value for the named header.
- Remove `getHeaderLines()` entirely.

At that point, the most common use cases are present, and the verbiage issues are fixed.

@pmjones Can you live with that? And are you understanding why I think `getHeaderLines()` should be removed, particularly with the link made above?

pjdietz commented:
I like removing `getHeaderLines()` as well as it doesn't really seem that useful and doesn't get you anything you can't get otherwise.

With that removed, this PR mostly just changes the name of the current `getHeaderLines($name)` to `getHeader($name)` and the current `getHeader($name)` to `getHeaderLine($name)`. It's a slight change, but seems a bit more consistent.

pmjones commented:
I can live with getting rid of getHeaderLines().

weierophinney commented:
@pjdietz Only thing remaining is to clarify where you have `@return string[]` on `getHeader()` to indicate "An array of discrete string values for the given header" or similar as a description of the return value. Otherwise :+1:.

pjdietz commented:
```
    /**
     * Retrieves a header by the given case-insensitive name as an array of
     * strings.
     *
     * If the header does not appear in the message, this method MUST return an
     * empty array.
     *
     * @param string $name Case-insensitive header field name.
     * @return string[] An array of discrete string values for the given header.
     */
    public function getHeader($name);
```
We have language above indicating an empty array for no header. Do you want to see that in the @return line as well?

pmjones commented:
FWIW #493 has been updated per this conversation.

weierophinney commented:
> We have language above indicating an empty array for no header. Do you want to see that in the @return line as well?

Yes, please.

pjdietz commented:
@weierophinney All set. I added corresponding language for `getHeaderLine()` as well.

weierophinney commented:
:ship:

mtdowling commented on proposed/http-message.md:
```diff
@@ -383,6 +383,20 @@ interface MessageInterface
+     * empty array.
+     *
+     * @param string $name Case-insensitive header field name.
+     * @return string[] An array of discrete string values for the given header.
```
I'm not sure what you mean by discrete here. I think it muddies the language unless clarified. For example, getHeader() should not parse each header in order to separate out CSV. If a user calls `setHeader('foo', ['bar, baz', 'bam'])`, then calling `getHeader('foo')` should return `['bar, baz', 'bam']`.

pjdietz commented:
@mtdowling Yes, `['bar, baz', 'bam']` is what I'd expect, too. I think discrete because you get an array item for each value. @weierophinney, was that your intent?

weierophinney commented:
@pjdietz and @mtdowling that was my intent. Perhaps the clarification should be "An array of string values as provided for the given header."

pjdietz commented:
Whoops, just saw the rest of your line. New one coming in a sec.

weierophinney commented:
:ship:

mtdowling commented:
:shipit: :boat:

weierophinney commented:
One last note (for @pjdietz and @pmjones): there are examples under section 1.2 that reference `getHeader()` and `getHeaderLines()` that need to be updated. Can one of you update those, or shall I open a separate pull request on merge?

pjdietz commented:
Sure, I can take care of it. Heading out now, but look for it in about an hour.

pmjones commented:
@pjdietz I have merged #493 as the basis for #494; please update from the new version, and then we can merge this PR on top of it.

pjdietz commented:
Ok, I've added a couple new commits. @pmjones, #493 was missing some of the return value descriptions worked out in the comments on this thread, so I added those back in. Also, section 1.2 should now be up to date.

weierophinney commented:
@pjdietz changes look good; thanks for making the changes to section 1.2!

@pmjones and/or @simensen — ready to merge!

pjdietz commented:
Glad to help!
